### ISSUE TITLE
Specify handling of multiple groups

### ISSUE THREAD

**Issue by ccrook** [opengeospatial/CRS-Gridded-Geodetic-data-eXchange-Format#53]
GGXF supports multiple groups.  Currently the specification provides no description of when they are used or how they are to be used.  The deformation model example includes multiple groups, and the [deformation model specification ](https://github.com/opengeospatial/CRS-Deformation-Models/raw/master/products/specification/abstract-specification-deformation-model-functional-model.pdf) describes how a deformation model contains multiple elements (eg secular deformation, coseismic deformation due to an earthquake), but we don't actually specify that the deformation model is encoded into GGXF by creating a ggxfGroup for each element of the model. 

Issue #52 has identified that the NADCON geographic3dOffset grids also require multiple groups, as the model has different grids (with different nodes) for the horizontal and vertical offsets.  This can be encoded into GGXF using two groups, but the GGXF specification doesn't define how these are to be used by software to compile the 3d offset.

The proof of concept has a mechanism for this that was designed for the deformation model, in which different groups may define a different subset of parameter values.  

For each group the code assumes that any parameters not explicitly defined by the model  are zero (that is parameters that are not defined by the ggxfGroup gridParameters attribute or constantParameters).  The interpolated parameter values and constant values are calculated for each group independently.  If the group includes a time function then that is evaluated and each group parameter value is added to define the final parameter value to be used in a coordinate operation.

The deformation model requires an additional feature for uncertainty parameters which is that total uncertainty is calculated as the root sum of squares of the values in each group.

To support this and ensure that the specification ensures that GGXF datasets with multiple groups are used in a well defined way to convey the producers intent it is propoosed that:

* the specification include an extra section "Multiple groups" , perhaps before 5.9 data extent, that defines how a user should use GGXF files with multiple groups.  This would describe the algorithm outlined above
* that an extra optional attribute _groupAdditionMethod_ is defined that may have values "addition" (the default) or "rootSumSquares".
* consideration be given to including another specification section "Time functions", possibly before Multiple Groups, that defines time functions and how they used, and C.3 Time functions defining their formulae.  They are defined in the deformation specification, but they may be valuable in the future for other data types (for example time varying geoid).  Including time functions and describing the algorithm above would allow the specification to stand as a complete document describing how GGXF data sets are used, and allow future requirements such as time varying geoid to be encoded without changing specification or software - they are already supported.

**Comment by ccrook** [comment]
@RogerLott FYI.  If you agree this is valuable I will draught the required sections.

**Comment by RogerLott** [comment]
@ccrook Yes, please, draft something. groupAdditionMethod and its possible values looks like an extension to the Conventions. Descriptive text might be an extension to 5.8.9 Parameters in a GGXF file containing multiple grids or might be included in commentary in the examples, or both.

Whilst issue #52 might have trigged this issue, in my view it is not directly related.  What is needed there (and needs to be put into issue #52) is for two offsets (latitudeOffset (δφ) and longitudeOffset (δλ)) to be derived through one ggxfGroup and a third offset (ellipsoidalHeightOffset (δh)) to be derived through a second ggxfGroup. All three offsets are declared through the file header, and the two ggxfGroup headers define the subset of these. GGXF handles this. What happens next is in my view nothing whatsoever to do with combining parameter values. Each of the three parameter values interpolated at a point are applied independently in the three formulae that the geographic3dOffset method requires. That would be true even if the parameters were held in a single ggxfGroup. It is a function of the method formula, not of the data stroage. 'combining' does not seem to me to be relevant; I am not sure that the issues are as closely related as you suggest.

**Comment by ccrook** [comment]
@RogerLott I agree it is debatable in the case of geographic3dOffsets since the groups define an discrete sets of parameters.  That is not the case in the deformation model, which is where I originally implemented this logic.  However as it generalises it seems cleaner to me to provide a common algorithm that allows more generous options in the future.

> Yes, please, draft something. groupAdditionMethod and its possible values looks like an extension to the Conventions. Descriptive text might be an extension to 5.8.9 Parameters in a GGXF file containing multiple grids or might be included in commentary in the examples, or both.

The groupAdditionMethod only applies where there are multiple groups.  within a single group there may be multiple grids, but only one is used.  I think our discussion of multiple grids is all in the context of a single group?

**Comment by ccrook** [comment]
@RogerLott.  There is one other feature relating to groups the comes primarily from deformation models.  That is whether the value of parameters outside the extents of the root grids in a group is zero, or whether it is no-data.  I'll draft something to cover that too.

**Comment by ccrook** [comment]
@RogerLott proposed text that may or may not warrant inclusion...
[MultipleGGXFGroups.docx](https://github.com/opengeospatial/CRS-Gridded-Geodetic-data-eXchange-Format/files/10189180/MultipleGGXFGroups.docx)

**Comment by ccrook** [comment]
Note that defining handling of multiple groups (in the same way as defining for multiple grids) provides useful potential capabilities for data types other than deformation model.

GGXF supports one way of refining a gridded function where a denser function is required by using multiple grids in a group.  However, only one of these grids is used at any location, so each grid must define the total value of that function.   That in turn places constraints on the densifying grid.  It must be aligned with the coarser grid and use the same interpolation method, as well as carefully calculated along the edge of the densifying grid, to have any chance of creating a continuous function.

Another valid way of densifying is to provide an additive "patch" grid that represents the local perturbation of the function from the coarser grid, so the coarse and fine grid values are added to determine the total field at any value.  This provides far fewer constraints in terms of the densifying grid.  It just needs to evaluate to zero at its edge for continuity.  Other than that it can have any grid spacing, any interpolation method, even use a rotated grid and the total function will still be continuous. 

Also calculating denser grid in a local region may often by done by modelling the different between the local data and what is predicted by the coarser grid, so this is a natural way to create to densify grids.  

For potentially inhomogeneous data types this seems like a very useful capability that is enabled by the two keywords _valueOutsideGrid_ and _groupAdditionMethod_. 

I see its potential value is well beyond deformation models.  It is not particularly difficult to implement - if we put this in the spec then compliant implementations will be expected to support it and producers can use it in the future without having to wait for new versions of GGXF spec and implementation.

A similar case can be made for time functions, although not as strong since a software for, say, applying geoid corrections, probably won't be written to provide a date (epoch or otherwise) to its geoid correction function.

**Comment by ccrook** [comment]
@RogerLott Reviewing this issue in light of the current state of GGXF.

There are two components to this issue.

### Handling of overlapping groups specifying the same (or different) parameters

GGXF includes the groupAdditionMethod which covers this.  The comment above "Alternative to specifying handling of multiple groups" doesn't apply.

### Application of time functions for types other than deformation

The argument was that a software that is implementing GGXF will be handling time functions for deformation, so why not simplify it by using them in the same way for all content types.  i.e. if an element includes a time function then the values are scaled by it.  

The counter-argument is that the time functions used for deformation may not be suitable for other content types, and also that it would create an awkward situation in terms of where the authoritative definitions of the time function formula lie.

The counter argument has prevailed!  We are only specifying time functions for deformation models.

Possibly the "timeFunctions" attribute should be renamed "deformationTimeFunction" or something like that in GGXF?
